### PR TITLE
java: add cloud functions test

### PR DESCRIPTION
**Pull request by daniel-sanche**
Replaces https://github.com/googleapis/env-tests-logging/pull/82

Implements Java Functions environment tests

Due to complications around Cloud Functions, this test uses a separate build method from the other deployment sources. It uses it's own `Dockerfile`, `pom.xml` file, and `CloudFunctionsTrigger.java` file. It shares the same `Snippets.java` file as the others though, so they will still be able to make use of a shared codebase

Implements https://github.com/googleapis/env-tests-logging/issues/53

### CODE REVIEW

**Review by minherz — COMMENTED** [review]
comprehensive work! my only concern is that i do not see any code that runs the client library in the CF handler. is it possible that you planned to do it with the next commit?
there are a few nit comments, mainly about the year in the copyright (we do not set past years to the new files).

File: deployable/java/functions/pom.xml [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        <groupId>com.google.cloud</groupId>
+        <artifactId>google-cloud-core</artifactId>
+        <version>2.8.3</version>
+    </dependency>
+    <dependency>
+        <groupId>com.google.cloud</groupId>
+        <artifactId>google-cloud-core-grpc</artifactId>
+        <version>2.8.3</version>
+    </dependency>
+  </dependencies>
```
[the remark this answers is not in the document]
Comment (daniel-sanche) [reply deployable/java/functions/pom.xml:62]: Yes, we do need these, or the library will throw errors. It looks like [renovate bot](https://github.com/googleapis/env-tests-logging/pull/83) is watching this repo, so they should stay up to date

File: deployable/java/functions/src/main/java/envtest/deployable/CloudFunctionTrigger.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    try {
+      Snippets obj = new Snippets();
+      Class<?> c = obj.getClass();
+      Method found = c.getDeclaredMethod(fnName, new Class[] {Map.class});
+      found.invoke(obj, args);
+    } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+      e.printStackTrace();
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (daniel-sanche) [reply deployable/java/functions/src/main/java/envtest/deployable/CloudFunctionTrigger.java:56]: The actual logs take place using the same [Snippets.java](https://github.com/googleapis/env-tests-logging/blob/main/deployable/java/src/main/java/envtest/deployable/Snippets.java) file as the other environments. I think it's important to keep all the actual logging logic in one place, so we can make sure we're testing the exact same code on multiple environments. The shared file is pulled in as part of the docker build phase [here](https://github.com/googleapis/env-tests-logging/blob/6fbb3da1e877a060b644c0617ed668247de90b16/deployable/java/Dockerfile.cloudfunctions#L53)

I'll admit I'm not super happy with how the files are organized. A part of me wants to find a way to merge all the java code into a single codebase, but that would add extra build complexity, and I'm not super familiar with maven. But let me know what you think

File: deployable/java/Dockerfile.cloudfunctions [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# compile local java-logging library
+FROM maven:3.8.4 AS lib-env
+WORKDIR /app
+COPY _library ./
+RUN mvn verify --fail-never
```
[the remark this answers is not in the document]
Comment (daniel-sanche) [reply deployable/java/Dockerfile.cloudfunctions]: good point, removed

File: deployable/java/Dockerfile.cloudfunctions [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+# compile local java-logging library
+FROM maven:3.8.4 AS lib-env
+WORKDIR /app
+COPY _library ./
+RUN mvn verify
+RUN mvn -Dmaven.test.skip=true package
+# copy the latest jar file to java-logging.jar
+RUN cp `ls google-cloud-logging/target/*.jar | grep -v "test" | sort -V | tail -n1` /app/java-logging.jar
+RUN cp `ls proto-google-cloud-logging-v2/target/*.jar | grep -v "test" | sort -V | tail -n1` /app/proto-java-logging.jar
```
Comment (minherz) [inline deployable/java/Dockerfile.cloudfunctions:23]: nit: consider combining multiple RUN commands to reduce the size of the image, hence improve building time. Same could be apply to the next image to improve uploading and launching times.

File: deployable/java/functions/src/main/java/envtest/deployable/CloudFunctionTrigger.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (minherz) [reply deployable/java/functions/src/main/java/envtest/deployable/CloudFunctionTrigger.java:55]: I agree. I think the problem is that Snippets.java is located in the different folder and it is hard to track/find it. Also, if the code is fixed, do we actually need a flexibility about the method name?
I think that the code will be easier to understand if we stick to one of the two approaches:
* everything is in one place (no use of Snippets.java)
* the call to "hard-coded" class and method are explicit and Snippets.java is easily discoverable

File: deployable/java/functions/src/main/java/envtest/deployable/CloudFunctionTrigger.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
+/*
+/* Copyright 2021 Google LLC
```
Comment (minherz) [inline deployable/java/functions/src/main/java/envtest/deployable/CloudFunctionTrigger.java:2]: nit: sorry for being picky, 2022 here too :slightly_smiling_face:

**Review by minherz — APPROVED** [review]
there are few more comments. they are mostly to improve the solution execution and readability. feel free to open separate issues to handle it later.

**Pull request merged**